### PR TITLE
Feature/approve certs

### PR DESCRIPTION
**Pull request by eheimbuch**
## Proposed changes

Add functions to approve unknown certificates and also reject them again.

### Your checklist for this pull request

**Contributor**:
- [x] PR is well described and the description can be used as a commit message on squash
- [x] Related issues linked to PR if existing and labels set
- [x] Target branch is not master (in most cases develop should bet the target of choice) 
- [x] Code does not conflict with target branch
- [x] New code is covered with unit tests
- [ ] New ui components are tested inside the storybook (module ui-components only) 
- [x] [Changelog entry file](https://github.com/scm-manager/changelog#changelog-entry-files) created in `gradle/changelog` or CHANGELOG.md is updated for plugins
- [x] Documentation updated (only necessary for new features or changed behaviour)

**Reviewer**:
- [ ] The clean code principles are respected ([CleanCode](https://clean-code-developer.com/virtues/))
- [ ] All new code/logic is implemented on the right spot / "Should this be done here?"
- [ ] UI changes fits into the layout
- [ ] The UI views / components are responsive (mobile views)
- [ ] Correct translations are available

### Checklist for branch merge request (not required for forks)

- [ ] Branch is green/blue on [Jenkins](https://oss.cloudogu.com/jenkins/)
- [ ] Quality Gate passed on [SonarQube](https://sonarcloud.io/organizations/scm-manager/projects)

### CODE REVIEW

File: src/main/java/com/cloudogu/sslcontext/CertificateException.java [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+import sonia.scm.ContextEntry;
+import sonia.scm.ExceptionWithContext;
+
+import java.util.List;
+
+public class CertificateException extends ExceptionWithContext {
+
+  public CertificateException(List<ContextEntry> context, String message, Exception cause) {
+    super(context, message, cause);
+  }
```
Comment (sdorra) [inline src/main/java/com/cloudogu/sslcontext/CertificateException.java:35]: The constructor is always called with `noContext`, so why is the context required?

File: src/main/java/com/cloudogu/sslcontext/Certificate.java [hunk 2]
```diff
@@ -89,6 +89,10 @@ public void reject() {
     }
   }
 
+  public void updateTimestamp() {
+    timestamp = Instant.now();
+  }
+
```
Comment (sdorra) [inline src/main/java/com/cloudogu/sslcontext/Certificate.java:95]: When must this method be called?

File: src/main/js/approved/ApprovedCertificateDetailsModal.tsx [hunk 3]
```diff
@@ -0,0 +1,175 @@
...
+  const [selectedCert, setSelectedCert] = useState<Certificate>(certificate);
+  const [error, setError] = useState<Error>();
+
+  const manageCertificate = (link: string) => {
+    apiClient
+      .post(link)
+      .then(() => onClose())
+      .then(() => refresh())
+      .catch(setError);
+  };
```
Comment (sdorra) [inline src/main/js/approved/ApprovedCertificateDetailsModal.tsx:91]: This function should be the return value of a custom hook and a loading state is missing.

File: src/main/js/approved/ApprovedCertificateDetailsModal.tsx [hunk 4]
```diff
@@ -0,0 +1,175 @@
...
+
+  const renderButton = () => {
+    if (!!selectedCert._links) {
+      if (selectedCert?._links.reject) {
+        return (
+          <Button
```
Comment (sdorra) [inline src/main/js/approved/ApprovedCertificateDetailsModal.tsx:97]: Missing loading state

File: src/main/js/rejected/RejectedCertificateDetailsModal.tsx [hunk 5]
```diff
@@ -0,0 +1,139 @@
...
+  const [selectedCert, setSelectedCert] = useState<Certificate>(certificate);
+  const [error, setError] = useState<Error>();
+
+  const manageCertificate = (link: string) => {
+    apiClient
+      .post(link)
+      .then(() => onClose())
+      .then(() => refresh())
+      .catch(setError);
+  };
```
Comment (sdorra) [inline src/main/js/rejected/RejectedCertificateDetailsModal.tsx:50]: Return of custom hook and missing loading state

File: src/main/js/rejected/RejectedCertificateDetailsModal.tsx [hunk 6]
```diff
@@ -0,0 +1,139 @@
...
+
+  const renderButton = () => {
+    if (!!selectedCert._links) {
+      if (selectedCert?._links.approve) {
+        return (
+          <Button
```
Comment (sdorra) [inline src/main/js/rejected/RejectedCertificateDetailsModal.tsx:56]: missing loading state

File: src/test/java/com/cloudogu/sslcontext/InMemoryBlobStore.java [hunk 7]
```diff
@@ -0,0 +1,132 @@
...
+import java.io.ByteArrayOutputStream;
+import java.io.FileNotFoundException;
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.OutputStream;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.UUID;
+
+public class InMemoryBlobStore implements BlobStore {
```
Comment (sdorra) [inline src/test/java/com/cloudogu/sslcontext/InMemoryBlobStore.java:40]: Is this a copy from scm-test?

File: src/test/java/com/cloudogu/sslcontext/InMemoryBlobStoreFactory.java [hunk 8]
```diff
@@ -0,0 +1,64 @@
...
+package com.cloudogu.sslcontext;
+
+import sonia.scm.store.BlobStore;
+import sonia.scm.store.BlobStoreFactory;
+import sonia.scm.store.StoreParameters;
+
+import java.util.HashMap;
+import java.util.Map;
+
+public class InMemoryBlobStoreFactory implements BlobStoreFactory {
```
Comment (sdorra) [inline src/test/java/com/cloudogu/sslcontext/InMemoryBlobStoreFactory.java:34]: Copy?

File: src/main/js/SSLContextOverview.tsx [hunk 9]
```diff
@@ -22,31 +22,34 @@
...
-    return <ErrorNotification error={error} />;
-  }
-
-  if (loading) {
-    return <Loading />;
-  }
+  const refreshAllData = () => {
+    rejectedResult.refresh();
+    approvedResult.refresh();
+  };
```
Comment (sdorra) [inline src/main/js/SSLContextOverview.tsx:45]: Do we need to refresh always both tables? It is not enough to refetch the approved table on approve and reject?

**Review by sdorra — CHANGES_REQUESTED** [review]

File: src/test/java/com/cloudogu/sslcontext/InMemoryBlobStore.java (hunk 7 above)
Comment (eheimbuch) [reply src/test/java/com/cloudogu/sslcontext/InMemoryBlobStore.java:40]: Yes, i depend on 2.0.0 which doesn't provide this test util classes

File: src/main/js/useManageCertificate.ts [hunk 10]
```diff
@@ -0,0 +1,43 @@
...
+  const [loading, setLoading] = useState(false);
+  const [error, setError] = useState<Error | undefined>();
+
+  const manage = (link: string) => {
+    setLoading(true);
+    apiClient
+      .post(link)
+      .then(() => refresh())
+      .then(() => setLoading(false))
+      .then(() => onClose())
```
Comment (sdorra) [inline src/main/js/useManageCertificate.ts:38]:
```suggestion
      .then(refresh)
      .then(() => setLoading(false))
      .then(onClose)
```

File: src/main/java/com/cloudogu/sslcontext/Certificate.java [hunk 11]
```diff
@@ -74,6 +74,7 @@ private String createFingerprint(byte[] certificate) {
   }
 
   public void approve() {
+    timestamp = Instant.now();
```
Comment (sdorra) [inline src/main/java/com/cloudogu/sslcontext/Certificate.java:77]: Why does the timestamp change if the status could not be changed? The timestamp should only updated, if the status was successfully updated.

File: src/main/java/com/cloudogu/sslcontext/Certificate.java [hunk 12]
```diff
@@ -82,6 +83,7 @@ public void approve() {
   }
 
   public void reject() {
+    timestamp = Instant.now();
```
Comment (sdorra) [inline src/main/java/com/cloudogu/sslcontext/Certificate.java:86]: Same as for approve

**Review by sdorra — CHANGES_REQUESTED** [review]

**Pull request merged**